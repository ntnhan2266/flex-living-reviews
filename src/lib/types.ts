export interface HostawayReview {
  id: number;
  type: string;
  status: string;
  rating: number | null;
  publicReview: string;
  reviewCategory: {
    category: string;
    rating: number;
  }[];
  submittedAt: string;
  guestName: string;
  listingName: string;
}

export type Channel = 'hostaway' | 'airbnb' | 'booking' | 'google';
export type ReviewStatus =
  | 'awaiting'
  | 'pending'
  | 'scheduled'
  | 'submitted'
  | 'published'
  | 'expired'
  | 'approved'
  | 'public';

export interface NormalizedReview {
  id: string;
  propertyName: string;
  guestName: string;
  rating: number;
  comment: string;
  date: Date;
  channel: Channel;
  categories: Record<string, number | undefined>;
  status: ReviewStatus;
}

export interface PropertyStats {
  id: string;
  name: string;
  averageRating: number;
  totalReviews: number;
  approvedReviews: number;
  recentTrend: 'up' | 'down' | 'stable';
}

export type CategoryKey =
  | 'cleanliness'
  | 'communication'
  | 'location'
  | 'accuracy'
  | 'checkin'
  | 'value';

export type SortKey = 'date_desc' | 'date_asc' | 'rating_desc' | 'rating_asc';

export type Filters = {
  property: string;
  channel: '' | 'hostaway' | 'airbnb' | 'booking' | 'google';
  rating: '' | '1' | '2' | '3' | '4' | '5';
  startDate: string; // yyyy-mm-dd
  endDate: string;   // yyyy-mm-dd
  status: 'all' | 'pending' | 'approved';
  category?: '' | CategoryKey;
  minCategory?: '' | '1' | '2' | '3' | '4' | '5';
  sort?: SortKey;
};

export interface ApiResponse<T> {
  status: 'success' | 'error';
  data?: T;
  message?: string;
  total?: number;
}
