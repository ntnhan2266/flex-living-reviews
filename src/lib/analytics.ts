// src/lib/analytics.ts
import { NormalizedReview, ReviewStatus } from '@/lib/types';

export type Granularity = 'day' | 'week' | 'month';

export type TimeseriesDatum = {
  date: string; // "YYYY-MM-DD" for day/week start/month start
  count: number; // number of reviews in bucket
  avgRating: number; // 0..5 average in bucket (0 when no ratings)
};

/**
 * Buckets a JS Date into an ISO key by granularity:
 * - day  -> YYYY-MM-DD
 * - week -> YYYY-MM-DD of the week's Monday
 * - month-> YYYY-MM-01
 */
export function bucketKey(d: Date, granularity: Granularity): string {
  const dt = new Date(d);
  dt.setHours(0, 0, 0, 0);

  if (granularity === 'week') {
    // Monday as week start
    const day = dt.getDay() || 7; // 1..7 (Mon..Sun)
    if (day !== 1) dt.setDate(dt.getDate() - (day - 1));
  } else if (granularity === 'month') {
    dt.setDate(1);
  }

  // fastest way to get YYYY-MM-DD without deps
  return dt.toISOString().slice(0, 10);
}

/**
 * Build a timeseries for reviews.
 * @param reviews - list of reviews
 * @param options.statuses - optional statuses to include (e.g. ['approved'])
 * @param options.granularity - day|week|month (default 'day')
 */
export function getReviewTimeseries(
  reviews: NormalizedReview[],
  options?: { statuses?: ReviewStatus[]; granularity?: Granularity },
): TimeseriesDatum[] {
  const allowed = new Set(options?.statuses ?? []);
  const gran = options?.granularity ?? 'day';

  const grouped: Record<string, { count: number; sum: number; withRating: number }> = {};

  for (const r of reviews) {
    if (allowed.size && !allowed.has(r.status)) continue;

    // normalize date (handle string | Date)
    const d = r.date instanceof Date ? r.date : new Date(r.date);
    const key = bucketKey(d, gran);
    if (!grouped[key]) grouped[key] = { count: 0, sum: 0, withRating: 0 };

    grouped[key].count += 1;

    const val = Number(r.rating);
    if (Number.isFinite(val)) {
      grouped[key].sum += val;
      grouped[key].withRating += 1;
    }
  }

  return Object.entries(grouped)
    .sort(([a], [b]) => (a < b ? -1 : a > b ? 1 : 0))
    .map(([date, { count, sum, withRating }]) => ({
      date,
      count,
      avgRating: withRating ? Number((sum / withRating).toFixed(2)) : 0,
    }));
}
