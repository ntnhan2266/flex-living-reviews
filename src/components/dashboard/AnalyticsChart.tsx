import React, { useMemo } from 'react';
import { NormalizedReview } from '@/lib/types';
import { getReviewTimeseries, TimeseriesDatum } from '@/lib/analytics';
import {
  ResponsiveContainer,
  ComposedChart,
  Line,
  Bar,
  XAxis,
  YAxis,
  Tooltip,
  CartesianGrid,
} from 'recharts';

type Props = {
  reviews: NormalizedReview[];
  granularity?: 'day' | 'week' | 'month';
  onlyApproved?: boolean; // default true: common KPI
  height?: number; // chart height
};

const DEFAULT_HEIGHT = 260;

const AnalyticsChart: React.FC<Props> = ({
  reviews,
  granularity = 'day',
  onlyApproved = true,
  height = DEFAULT_HEIGHT,
}) => {
  const data: TimeseriesDatum[] = useMemo(() => {
    if (!reviews?.length) return [];
    return getReviewTimeseries(reviews, {
      granularity,
      statuses: onlyApproved ? ['approved'] : undefined,
    });
  }, [reviews, granularity, onlyApproved]);

  if (!data.length) {
    return (
      <div className="bg-white p-6 rounded-lg shadow text-center text-gray-900">
        No analytics data available.
      </div>
    );
  }

  return (
    <div className="bg-white p-6 rounded-lg shadow">
      <div className="flex items-baseline justify-between mb-4">
        <h2 className="text-lg font-semibold text-gray-800">Review Analytics</h2>
        <div className="text-xs text-gray-500 uppercase tracking-wide">
          {onlyApproved ? 'Approved' : 'All'} Â· {granularity}
        </div>
      </div>

      <ResponsiveContainer width="100%" height={height}>
        <ComposedChart data={data} margin={{ top: 10, right: 20, left: 0, bottom: 0 }}>
          <CartesianGrid strokeDasharray="3 3" />
          <XAxis dataKey="date" tick={{ fontSize: 12 }} />
          <YAxis yAxisId="left" domain={[0, 5]} allowDecimals={false} />
          <YAxis yAxisId="right" orientation="right" allowDecimals={false} />
          <Tooltip
            formatter={(value: number, name: string) => {
              if (name === 'avgRating') return [Number(value).toFixed(2), 'Avg Rating'];
              return [value, 'Count'];
            }}
            labelFormatter={(label: string) => `Date: ${label}`}
          />
          <Bar
            yAxisId="right"
            dataKey="count"
            name="Count"
            barSize={18}
            fill="#93c5fd" /* Tailwind blue-300 */
          />
          <Line
            yAxisId="left"
            type="monotone"
            dataKey="avgRating"
            name="Avg Rating"
            stroke="#2563eb" /* Tailwind blue-600 */
            strokeWidth={2}
            dot={false}
          />
        </ComposedChart>
      </ResponsiveContainer>
    </div>
  );
};

export default AnalyticsChart;
