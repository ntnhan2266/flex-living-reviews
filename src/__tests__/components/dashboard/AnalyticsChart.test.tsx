import { render, screen } from '@testing-library/react';
import AnalyticsChart from '@/components/dashboard/AnalyticsChart';
import { NormalizedReview } from '@/lib/types';

const base: NormalizedReview = {
  id: '1',
  propertyName: 'Flat A',
  guestName: 'Jane Doe',
  rating: 4,
  comment: 'Great stay',
  date: new Date('2025-01-08'),
  channel: 'hostaway',
  categories: {},
  status: 'approved',
};

describe('<AnalyticsChart />', () => {
  it('renders empty state when no reviews', () => {
    render(<AnalyticsChart reviews={[]} />);
    expect(screen.getByText(/No analytics data available/i)).toBeInTheDocument();
  });

  it('renders chart title when reviews are provided', () => {
    render(<AnalyticsChart reviews={[base]} />);
    expect(screen.getByText(/Review Analytics/i)).toBeInTheDocument();
  });

  it('shows granularity options', () => {
    render(<AnalyticsChart reviews={[base]} />);
    expect(screen.getByText(/DAY/i)).toBeInTheDocument();
  });

  it('matches snapshot with reviews', () => {
    const { asFragment } = render(<AnalyticsChart reviews={[base]} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('matches snapshot without reviews', () => {
    const { asFragment } = render(<AnalyticsChart reviews={[]} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
