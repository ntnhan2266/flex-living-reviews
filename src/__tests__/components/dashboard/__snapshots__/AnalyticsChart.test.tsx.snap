// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<AnalyticsChart /> matches snapshot with reviews 1`] = `
<DocumentFragment>
  <div
    class="bg-white p-6 rounded-lg shadow"
  >
    <div
      class="flex items-baseline justify-between mb-4"
    >
      <h2
        class="text-lg font-semibold text-gray-800"
      >
        Review Analytics
      </h2>
      <div
        class="text-xs text-gray-500 uppercase tracking-wide"
      >
        Approved Â· day
      </div>
    </div>
    <div
      class="recharts-responsive-container"
      style="width: 100%; height: 260px; min-width: 0;"
    >
      <div
        style="width: 0px; height: 0px; overflow: visible;"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`<AnalyticsChart /> matches snapshot without reviews 1`] = `
<DocumentFragment>
  <div
    class="bg-white p-6 rounded-lg shadow text-center text-gray-900"
  >
    No analytics data available.
  </div>
</DocumentFragment>
`;
